src/handlers/VehicleTestController.ts

```diff
@@ -1,58 +1,230 @@
+        payloadWithCertificateNumber
+      );
+      payloadWithAnniversaryDate.vehicleId = payloadWithAnniversaryDate.vrm;
+      const result = await this.dataProvider.insertTestResult(payloadWithAnniversaryDate);
+      return result;
+    } catch (error) {
+      if (error.statusCode === 400 && error.message === enums.MESSAGES.CONDITIONAL_REQUEST_FAILED) {
+          console.info(
+            "TestResultService.insertTestResult: Test Result id already exists",
+            error
+          );
+          error =  new models.HTTPResponse(201, enums.MESSAGES.ID_ALREADY_EXISTS);
+        }
+      throw error;
+    }
+  }
+//#endregion
+//#region [rgba(0, 205, 30, 0.1)] Private functions
+  /**
+   * Note: When performing actions on a moment instance, it gets mutated
+   * Note: Expiry dates on the payload should be set at the UTC start of day.
+   *
+   * @param payload
+   */
+  private async generateExpiryDate(
+    payload: models.ITestResultPayload
+  ): Promise<models.ITestResultPayload> {
+    try {
+      if (
+        payload.testStatus !== enums.TEST_STATUS.SUBMITTED ||
+        utils.ValidationUtil.isNotAllowedVehicleTypeForExpiry(
+          payload.vehicleType
+        )
+      ) {
+        return Promise.resolve(payload);
+      }
+      const expiryTestTypes = payload.testTypes.filter((testType) =>
+        utils.ValidationUtil.isAllowedTestTypeForExpiry(testType)
+      );
+
+      const recentExpiry = await this.dataProvider.getMostRecentExpiryDate(
+        payload.systemNumber
+      );
+
+      expiryTestTypes.forEach((testType: any, index: number) => {
+        const testTypeForExpiry: models.TestTypeForExpiry = {
+          testType,
+          vehicleType:
+            enums.VEHICLE_TYPE[
+              payload.vehicleType.toUpperCase() as keyof typeof enums.VEHICLE_TYPE
+            ],
+          recentExpiry,
+          regnOrFirstUseDate: VehicleTestController.getRegistrationOrFirstUseDate(payload),
+          hasHistory: !DateProvider.isSameAsEpoc(recentExpiry),
+          hasRegistration: DateProvider.isValidDate(
+            VehicleTestController.getRegistrationOrFirstUseDate(payload)
+          ),
+        };
+        console.log(testTypeForExpiry);
+        const strategy = this.getExpiryStrategy(testTypeForExpiry);
+        console.log(strategy.constructor.name);
+        testType.testExpiryDate = strategy.getExpiryDate();
+      });
+      console.log("generateExpiryDate: testTypes ->", payload.testTypes);
+      return Promise.resolve(payload);
+    } catch (error) {
+      console.error("Error in error generateExpiryDate", error);
+      throw new models.HTTPError(500, enums.MESSAGES.INTERNAL_SERVER_ERROR);
+    }
+  }
+
+  private static AssignCertificateNumberToTestTypes(
+    payload: models.ITestResultPayload
+  ) {
+    if (payload.testStatus !== enums.TEST_STATUS.SUBMITTED) {
+      return payload;
+    }
+    const { vehicleType } = payload;
```

Comment:
we could also destructure testTypes here if we wanted

Reply:
We are doing a for each for testType to assign testNumber to certificateNumber.